File: backend/app/schemas/review.py
"""Wrong-answer notebook schemas.

These models define the strict API contract for grouped review data.
"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


class WrongAnswerQuestion(BaseModel):
    """Single wrong-answer record used inside a node group."""

    question_record_id: str = Field(..., description="Unique answer record ID")
    question_id: Optional[str] = Field(default=None, description="Original question ID")
    node_id: str = Field(..., description="Knowledge node ID")
    question_text: str = Field(..., description="Question text")
    user_answer: str = Field(..., description="Student answer")
    correct_answer: str = Field(..., description="Correct answer")
    error_type: str = Field(..., description="Validation error classification")
    error_severity: int = Field(..., ge=1, le=3, description="Error severity 1-3")
    question_type: Literal["multiple_choice", "short_answer"] = Field(
        default="multiple_choice",
        description="Question type",
    )
    attempted_at: str = Field(..., description="ISO 8601 timestamp")
    is_invalidated: bool = Field(default=False, description="Hallucination flag")


class WrongAnswerNodeGroup(BaseModel):
    """Grouped wrong answers for one knowledge node."""

    node_id: str = Field(..., description="Knowledge node ID")
    node_label: str = Field(..., description="Knowledge node label")
    total_errors: int = Field(..., ge=0, description="Number of wrong answers")
    questions: list[WrongAnswerQuestion] = Field(default_factory=list)


class WrongAnswersSummary(BaseModel):
    """Aggregate wrong-answer summary."""

    total_wrong_count: int = Field(..., ge=0)
    total_nodes_with_errors: int = Field(..., ge=0)


class WrongAnswersData(BaseModel):
    """Data payload for the wrong-answer notebook."""

    by_node: list[WrongAnswerNodeGroup] = Field(default_factory=list)
    summary: WrongAnswersSummary


class WrongAnswersResponse(BaseModel):
    """Standard response envelope for the review API."""

    status: Literal["success", "error"] = Field(default="success")
    data: Optional[WrongAnswersData] = Field(default=None)
    message: str = Field(default="")
    trace_id: str = Field(default="")


class ChapterMasteryItem(BaseModel):
    """Mastery stats grouped by one parent chapter."""

    parent_id: str = Field(..., description="Parent chapter node ID")
    parent_label: str = Field(..., description="Parent chapter label")
    attempted_count: int = Field(..., ge=0)
    correct_count: int = Field(..., ge=0)
    mastery_score: float = Field(..., ge=0.0, le=1.0)


class ChapterMasterySummary(BaseModel):
    """Aggregate summary for chapter mastery."""

    total_parents: int = Field(..., ge=0)
    total_attempted: int = Field(..., ge=0)
    total_correct: int = Field(..., ge=0)
    overall_mastery_score: float = Field(..., ge=0.0, le=1.0)


class ChapterMasteryData(BaseModel):
    """Data payload for chapter mastery metrics."""

    by_parent: list[ChapterMasteryItem] = Field(default_factory=list)
    summary: ChapterMasterySummary


class ChapterMasteryResponse(BaseModel):
    """Standard response envelope for chapter mastery API."""

    status: Literal["success", "error"] = Field(default="success")
    data: Optional[ChapterMasteryData] = Field(default=None)
    message: str = Field(default="")
    trace_id: str = Field(default="")


class InvalidateQuestionRequest(BaseModel):
    """Request payload to invalidate one question record."""

    question_record_id: str = Field(..., min_length=1)
    reason: Optional[str] = Field(default=None, max_length=500)


class InvalidateQuestionData(BaseModel):
    """Mutation result details for an invalidation request."""

    question_record_id: str = Field(...)
    found: bool = Field(...)
    updated: bool = Field(...)
    already_invalidated: bool = Field(...)
    invalidated_at: Optional[str] = Field(default=None)


class InvalidateQuestionResponse(BaseModel):
    """Standard response envelope for question invalidation."""

    status: Literal["success", "error"] = Field(default="success")
    data: Optional[InvalidateQuestionData] = Field(default=None)
    message: str = Field(default="")
    trace_id: str = Field(default="")

File: frontend/src/api/review.ts
import type { ApiEnvelope } from '../types'

const API_BASE = '/api/v1'
const USER_ID_STORAGE_KEY = 'ai-class-user-id'

export interface WrongAnswerQuestion {
  question_record_id: string
  question_id: string | null
  node_id: string
  question_text: string
  user_answer: string
  correct_answer: string
  error_type: string
  error_severity: number
  question_type: 'multiple_choice' | 'short_answer'
  attempted_at: string
  is_invalidated: boolean
}

export interface WrongAnswerNodeGroup {
  node_id: string
  node_label: string
  total_errors: number
  questions: WrongAnswerQuestion[]
}

export interface WrongAnswersData {
  by_node: WrongAnswerNodeGroup[]
  summary: {
    total_wrong_count: number
    total_nodes_with_errors: number
  }
}

export interface ChapterMasteryItem {
  parent_id: string
  parent_label: string
  attempted_count: number
  correct_count: number
  mastery_score: number
}

export interface ChapterMasteryData {
  by_parent: ChapterMasteryItem[]
  summary: {
    total_parents: number
    total_attempted: number
    total_correct: number
    overall_mastery_score: number
  }
}

export interface InvalidateQuestionRequest {
  question_record_id: string
  reason?: string
}

export interface InvalidateQuestionData {
  question_record_id: string
  found: boolean
  updated: boolean
  already_invalidated: boolean
  invalidated_at: string | null
}

function getClientUserId(): string {
  if (typeof window === 'undefined') {
    return 'anonymous'
  }

  const storedUserId = window.localStorage.getItem(USER_ID_STORAGE_KEY)
  if (storedUserId && storedUserId.trim()) {
    return storedUserId
  }

  const generatedUserId =
    typeof crypto !== 'undefined' && 'randomUUID' in crypto
      ? crypto.randomUUID()
      : `user-${Date.now()}-${Math.random().toString(16).slice(2)}`

  window.localStorage.setItem(USER_ID_STORAGE_KEY, generatedUserId)
  return generatedUserId
}

export async function getWrongAnswers(
  nodeIdFilter?: string
): Promise<ApiEnvelope<WrongAnswersData>> {
  const url = new URL(`${API_BASE}/review/wrong-answers`, window.location.origin)
  if (nodeIdFilter && nodeIdFilter.trim()) {
    url.searchParams.set('node_id_filter', nodeIdFilter.trim())
  }

  const response = await fetch(url.toString(), {
    headers: {
      'X-User-ID': getClientUserId(),
    },
  })

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    const errorMessage = errorData?.message || `API error: ${response.status}`
    throw new Error(errorMessage)
  }

  return response.json()
}

export async function getChapterMastery(): Promise<ApiEnvelope<ChapterMasteryData>> {
  const url = new URL(`${API_BASE}/review/chapter-mastery`, window.location.origin)

  const response = await fetch(url.toString(), {
    headers: {
      'X-User-ID': getClientUserId(),
    },
  })

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    const errorMessage = errorData?.message || `API error: ${response.status}`
    throw new Error(errorMessage)
  }

  return response.json()
}

export async function invalidateQuestionRecord(
  payload: InvalidateQuestionRequest
): Promise<ApiEnvelope<InvalidateQuestionData>> {
  const response = await fetch(`${API_BASE}/review/invalidate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-User-ID': getClientUserId(),
    },
    body: JSON.stringify(payload),
  })

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    const errorMessage = errorData?.message || `API error: ${response.status}`
    throw new Error(errorMessage)
  }

  return response.json()
}
